### PR TITLE
Update selector parser

### PR DESCRIPTION
**Pull request by jeddy3**
> Which issue, if any, is this issue related to?

References https://github.com/orgs/stylelint/teams/core/discussions/18

Replaces https://github.com/stylelint/stylelint/pull/3988

> Is there anything in the PR that needs further explanation?

I've ported over the changes from https://github.com/stylelint/stylelint/pull/3988. We changed our formatting and moved to TS JS Docs from flow since we made the original pull request

I believe only two rules are affected by this update:

- `selector-descendant-combinator-no-non-space` will now outright ignore selectors containing comments
- `selector-pseudo-class-parentheses-space-inside` can now longer autofix pseudo-classes that contain comments

I think these are minor limitations for the sake of updating. I've added a note to the READMEs of both these rules.

We have fixed some other bugs by updating the selector parser, e.g. the one in [`selector-attribute-brackets-space-inside/__tests__/index.js`](https://github.com/stylelint/stylelint/compare/update-selector-parser?expand=1#diff-0e3635f410297f27f1bfa58ad592f9a4). Things probably balance out in the end.

I've skipped the affected tests.

---

It'd be great if we could get https://github.com/stylelint/stylelint/pull/4592 and https://github.com/stylelint/stylelint/pull/4591 approved and merged too so that the new workaround feature and related documentation changes are also released.

### CODE REVIEW

File: lib/utils/isStandardSyntaxCombinator.js [hunk 1]
```diff
@@ -10,23 +10,5 @@ module.exports = function(node) {
 	// Ghost descendant combinators around reference combinators like `/deep/`
 	// postcss-selector-parser parsers references combinators as tag selectors surrounded
```
Comment (hudochenkov) [inline lib/utils/isStandardSyntaxCombinator.js]: This comment shall be changed to match new parser.

File: lib/rules/string-quotes/index.js [hunk 2]
```diff
@@ -72,55 +72,86 @@ function rule(expectation, secondary, context) {
...
+						return;
+					}
 
-						if (avoidEscape && needsEscape) {
-							// don't consider this an error
-							return;
+					if (attributeNode.quoteMark === correctQuote) {
+						if (avoidEscape) {
+							const needsCorrectEscape = attributeNode.value.indexOf(correctQuote) !== -1;
+							const needsOtherEscape = attributeNode.value.indexOf(erroneousQuote) !== -1;
```
Comment (hudochenkov) [inline lib/rules/string-quotes/index.js]: A little bit of readability:

```js
const needsCorrectEscape = attributeNode.value.includes(correctQuote);
const needsOtherEscape = attributeNode.value.includes(erroneousQuote);
```

File: lib/rules/string-quotes/index.js [hunk 3]
```diff
@@ -72,55 +72,86 @@ function rule(expectation, secondary, context) {
...
-								// the length of our value
-								attributeNode.value.length -
-								// with the length of our quote subtracted
-								erroneousQuote.length;
-
-							fixPositions.push(openIndex, closeIndex);
+					if (attributeNode.quoteMark === erroneousQuote) {
+						if (avoidEscape) {
+							const needsCorrectEscape = attributeNode.value.indexOf(correctQuote) !== -1;
+							const needsOtherEscape = attributeNode.value.indexOf(erroneousQuote) !== -1;
```
Comment (hudochenkov) [inline lib/rules/string-quotes/index.js]:
```js
const needsCorrectEscape = attributeNode.value.includes(correctQuote);
const needsOtherEscape = attributeNode.value.includes(erroneousQuote);
```

**Comment by jeddy3** [comment]
Good catches. I've made the changes.

**Comment by hudochenkov** [comment]
Do you think we should mention in a changelog that some rules were downgraded?

**Comment by jeddy3** [comment]
> - Security: updated to `postcss-selector-parser@6` due to a vulnerability in one of `postcss-selector-parser@3` dependencies ([#4595](https://github.com/stylelint/stylelint/pull/4595)).

**Comment by jeddy3** [comment]
> Do you think we should mention in a changelog that some rules were downgraded?

I think so.

**Comment by jeddy3** [comment]
> - Security: updated to `postcss-selector-parser@6` due to a vulnerability in one of `postcss-selector-parser@3` dependencies.([#4595](https://github.com/stylelint/stylelint/pull/4595)). Due to this update: 
>   - `selector-descendant-combinator-no-non-space` will ignore selectors containing comments
>   - `selector-pseudo-class-parentheses-space-inside` can't autofix pseudo-classes that contain comments

**Comment by jeddy3** [comment]
> We shall create an issue with all of these downgrades. To remember to restore functionality when time comes.

Done in https://github.com/stylelint/stylelint/issues/4599 & https://github.com/stylelint/stylelint/issues/4600.

**Pull request merged**